src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php

```diff
@@ -162,176 +156,452 @@ private function compileRoutes(RouteCollection $routes, $supportsRedirections)
+            $hasTrailingSlash = $hasTrailingSlash && (!$methods || isset($methods['HEAD']) || isset($methods['GET']));
+            $state->mark += 3 + $state->markTail + $hasTrailingSlash + strlen($regex) - $prefixLen;
+            $state->markTail = 2 + strlen($state->mark);
+            $code .= "\n                    .";
+            $code .= self::export(sprintf('|%s(*:%s)', substr($regex, $prefixLen).($hasTrailingSlash ? '?' : ''), $state->mark));
+            $vars = array_merge($state->hostVars, $vars);
+
+            if (!$route->getCondition() && (!is_array($next = $routes[1 + $i] ?? null) || $regex !== $next[1])) {
+                $prevRegex = null;
+                $state->checkTrailingSlash = $state->checkTrailingSlash || $hasTrailingSlash;
+                $state->default .= sprintf(
+                    "%s => array(%s, %s, %s, %s),\n",
+                    $state->mark,
+                    self::export(array('_route' => $name) + $route->getDefaults()),
+                    self::export($vars),
+                    self::export($methods ?: null),
+                    self::export(array_flip($route->getSchemes()) ?: null).($hasTrailingSlash ? ', true' : '')
+                );
+            } else {
+                $prevRegex = $compiledRoute->getRegex();
+                $combine = '            $matches = array(';
+                foreach ($vars as $j => $m) {
+                    $combine .= sprintf('%s => $matches[%d] ?? null, ', self::export($m), 1 + $j);
+                }
+                $combine = $vars ? substr_replace($combine, ");\n\n", -2) : '';
+
+                $state->switch .= <<<EOF
+        case {$state->mark}:
+{$combine}{$this->compileRoute($route, $name, $state->supportsRedirections, $hasTrailingSlash)}
+            break;
 
 EOF;
-                } else {
-                    $code .= <<<EOF
-            if ('$methods[0]' !== \$canonicalMethod) {
-                \$allow[] = '$methods[0]';
-                goto $gotoname;
             }
+        }
 
+        return $code;
+    }
+
+    /**
+     * A simple helper to compiles the switch's "default" for both static and dynamic routes.
+     */
+    private function compileSwitchDefault(bool $hasVars, string $routesKey, bool $matchHost, bool $supportsRedirections, bool $checkTrailingSlash)
+    {
+        if ($hasVars) {
+            $code = <<<EOF
+
+            foreach (\$vars as \$i => \$v) {
+                if (isset(\$matches[1 + \$i])) {
+                    \$ret[\$v] = \$matches[1 + \$i];
+                }
+            }
 
 EOF;
+        } elseif ($matchHost) {
+            $code = <<<EOF
+
+            if (\$requiredHost) {
+                if ('#' !== \$requiredHost[0] ? \$requiredHost !== \$host : !preg_match(\$requiredHost, \$host, \$hostMatches)) {
+                    break;
                 }
+                if ('#' === \$requiredHost[0] && \$hostMatches) {
+                    \$hostMatches['_route'] = \$ret['_route'];
```

Comment:
Can't this be removed? It's only when the host has a `_route` placeholder. This does not seem to be a supported case. You can only get the route name or the content of the placeholder. But not both. So using _route name should be forbidden as placeholder.

Reply:
> using _route name should be forbidden as placeholder

It should, but it isn't, and the current has actual protection so that the route name always has precedence. Looking twice, I think this is required yes.

Reply:
we should deprecate defining such placeholder though, to avoid WTF